diff --git a/src/main/sh/mommy b/src/main/sh/mommy
index 41abe71..accddc0 100755
--- a/src/main/sh/mommy
+++ b/src/main/sh/mommy
@@ -6,6 +6,8 @@ n="
 MOMMY_CAREGIVER="mommy"
 MOMMY_PRONOUNS="she her her"
 MOMMY_SWEETIE="${LOGNAME:-girl}"
+MOMMY_PET="slut/toy/pet/pervert/whore"
+MOMMY_PART="milk"
 MOMMY_PREFIX=""
 MOMMY_SUFFIX="~"
 
@@ -281,14 +283,18 @@ split_pronouns() {
 # 5. stores the output in `$fill_template__out`.
 fill_template() {
     list_normal_choose "$2"; sweetie="$list_choose__out"
-    list_normal_choose "$3"; split_pronouns "$list_choose__out"
-    list_normal_choose "$4"; caregiver="$list_choose__out"
+    list_normal_choose "$3"; pet="$list_choose__out"
+    list_normal_choose "$4"; part="$list_choose__out"
+    list_normal_choose "$5"; split_pronouns "$list_choose__out"
+    list_normal_choose "$6"; caregiver="$list_choose__out"
 
-    list_normal_choose "$5"; prefix="$list_choose__out"
-    list_normal_choose "$6"; suffix="$list_choose__out"
+    list_normal_choose "$7"; prefix="$list_choose__out"
+    list_normal_choose "$8"; suffix="$list_choose__out"
 
     replace_all__out="$1"
     replace_all "$replace_all__out" "%%SWEETIE%%" "$sweetie"
+    replace_all "$replace_all__out" "%%PET%%" "$pet"
+    replace_all "$replace_all__out" "%%PART%%" "$part"
     replace_all "$replace_all__out" "%%THEY%%" "$they"
     replace_all "$replace_all__out" "%%THEM%%" "$them"
     replace_all "$replace_all__out" "%%THEIR%%" "$their"
@@ -296,7 +302,7 @@ fill_template() {
     replace_all "$replace_all__out" "%%N%%" "$n"
     replace_all "$replace_all__out" "%%S%%" "/"
 
-    capitalize "$prefix$replace_all__out$suffix" "$7"; fill_template__out="$capitalize__out"
+    capitalize "$prefix$replace_all__out$suffix" "$9"; fill_template__out="$capitalize__out"
     return 0
 }
 
@@ -398,7 +404,7 @@ else
     # Select and fill template
     list_normalize "$MOMMY_FORBIDDEN_WORDS"; forbidden_words="$list_normalize__out"
     list_normal_choose "$templates" "$forbidden_words"; template="$list_choose__out"
-    fill_template "$template" "$MOMMY_SWEETIE" "$MOMMY_PRONOUNS" "$MOMMY_CAREGIVER" "$MOMMY_PREFIX" "$MOMMY_SUFFIX" \
+    fill_template "$template" "$MOMMY_SWEETIE" "$MOMMY_PET" "$MOMMY_PART" "$MOMMY_PRONOUNS" "$MOMMY_CAREGIVER" "$MOMMY_PREFIX" "$MOMMY_SUFFIX" \
                   "$MOMMY_CAPITALIZE"; response="$fill_template__out"
     list_filter_not "$response" "$forbidden_words"; safe_response="$list_filter_not__out"
 
